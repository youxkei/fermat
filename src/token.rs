#[cfg(test)]
use insta::assert_debug_snapshot_matches;

#[derive(Clone, Copy, Debug)]
pub struct Position {
    pub offset: usize,
    pub line: usize,
    pub column: usize,
}

impl Position {
    fn forward(&mut self, n: usize) {
        self.offset += n;
        self.column += n;
    }

    fn break_line(&mut self) {
        self.column = 1;
        self.line += 1;
    }
}

#[derive(Debug)]
pub enum Symbol {
    OpenParen,
    CloseParen,
    Comma,
    HyphenGreater,
    OpenBrace,
    CloseBrace,
    OpenBracket,
    CloseBracket,
    Bar,
    BarBar,
    LessHyphen,
    Semicolon,
    Colon,
    NumberSign,
    Dot,
    Asterisk,
    Slash,
    Plus,
    Hyphen,
    PlusPlus,
    HyphenHyphen,
    EqualEqual,
    SlashEqual,
    EqualLess,
    Less,
    GreaterEqual,
    Greater,
    EqualColonEqual,
    EqualSlashEqual,
    LessEqual,
    EqualGreater,
    ColonEqual,
    LessLess,
    GreaterGreater,
    Exclamation,
    Equal,
    ColonColon,
    DotDot,
    DotDotDot,
    Question,
    NewLine,
}

#[derive(Debug)]
pub enum Keyword {
    After,
    Begin,
    Case,
    Try,
    Catch,
    End,
    Fun,
    If,
    Of,
    Receive,
    When,
    Andalso,
    Orelse,
    Bnot,
    Not,
    Div,
    Rem,
    Band,
    And,
    Bor,
    Bxor,
    Bsl,
    Bsr,
    Or,
    Xor,
}

#[derive(Debug)]
pub enum TokenType {
    Integer,
    Float,
    Atom,
    String,
    Variable,
    Symbol(Symbol),
    Keyword(Keyword),
    Comment,
    Whitespace,
}

#[derive(Debug)]
pub struct Token<'a> {
    pub text: &'a [u8],
    pub position: Position,
    pub token_type: TokenType,
}

#[derive(Debug)]
pub enum Error {}

pub struct Tokenizer<'a> {
    src: &'a [u8],
    current_position: Position,
}

impl<'a> Tokenizer<'a> {
    pub fn new(src: &'a [u8]) -> Self {
        Tokenizer {
            src,
            current_position: Position { offset: 0, line: 1, column: 1 },
        }
    }

    fn tokenize_symbol(&mut self, length: usize, symbol: Symbol, rest: &'a [u8]) -> Option<Result<Token<'a>, Error>> {
        let src = self.src;
        let position = self.current_position;

        self.src = rest;
        self.current_position.forward(length);
        if let Symbol::NewLine = symbol {
            self.current_position.break_line();
        }

        Some(Ok(Token {
            text: &src[0..length],
            position,
            token_type: TokenType::Symbol(symbol),
        }))
    }
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Result<Token<'a>, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            match self.src {
                &[] => return None,

                &[b'=', b':', b'=', ref rest..] => return self.tokenize_symbol(3, Symbol::EqualColonEqual, rest),
                &[b'=', b'/', b'=', ref rest..] => return self.tokenize_symbol(3, Symbol::EqualSlashEqual, rest),
                &[b'.', b'.', b'.', ref rest..] => return self.tokenize_symbol(3, Symbol::DotDotDot, rest),

                &[b'-', b'>', ref rest..] => return self.tokenize_symbol(2, Symbol::HyphenGreater, rest),
                &[b'|', b'|', ref rest..] => return self.tokenize_symbol(2, Symbol::BarBar, rest),
                &[b'<', b'-', ref rest..] => return self.tokenize_symbol(2, Symbol::LessHyphen, rest),
                &[b'+', b'+', ref rest..] => return self.tokenize_symbol(2, Symbol::PlusPlus, rest),
                &[b'-', b'-', ref rest..] => return self.tokenize_symbol(2, Symbol::HyphenHyphen, rest),
                &[b'=', b'=', ref rest..] => return self.tokenize_symbol(2, Symbol::EqualEqual, rest),
                &[b'/', b'=', ref rest..] => return self.tokenize_symbol(2, Symbol::SlashEqual, rest),
                &[b'=', b'<', ref rest..] => return self.tokenize_symbol(2, Symbol::EqualLess, rest),
                &[b'>', b'=', ref rest..] => return self.tokenize_symbol(2, Symbol::GreaterEqual, rest),
                &[b'<', b'=', ref rest..] => return self.tokenize_symbol(2, Symbol::LessEqual, rest),
                &[b'=', b'>', ref rest..] => return self.tokenize_symbol(2, Symbol::EqualGreater, rest),
                &[b':', b'=', ref rest..] => return self.tokenize_symbol(2, Symbol::ColonEqual, rest),
                &[b'<', b'<', ref rest..] => return self.tokenize_symbol(2, Symbol::LessLess, rest),
                &[b'>', b'>', ref rest..] => return self.tokenize_symbol(2, Symbol::GreaterGreater, rest),
                &[b':', b':', ref rest..] => return self.tokenize_symbol(2, Symbol::ColonColon, rest),
                &[b'.', b'.', ref rest..] => return self.tokenize_symbol(2, Symbol::DotDot, rest),
                &[b'\r', b'\n', ref rest..] => return self.tokenize_symbol(2, Symbol::NewLine, rest),

                &[b'(', ref rest..] => return self.tokenize_symbol(1, Symbol::OpenParen, rest),
                &[b')', ref rest..] => return self.tokenize_symbol(1, Symbol::CloseParen, rest),
                &[b',', ref rest..] => return self.tokenize_symbol(1, Symbol::Comma, rest),
                &[b'{', ref rest..] => return self.tokenize_symbol(1, Symbol::OpenBrace, rest),
                &[b'}', ref rest..] => return self.tokenize_symbol(1, Symbol::CloseBrace, rest),
                &[b'[', ref rest..] => return self.tokenize_symbol(1, Symbol::OpenBracket, rest),
                &[b']', ref rest..] => return self.tokenize_symbol(1, Symbol::CloseBracket, rest),
                &[b'|', ref rest..] => return self.tokenize_symbol(1, Symbol::Bar, rest),
                &[b';', ref rest..] => return self.tokenize_symbol(1, Symbol::Semicolon, rest),
                &[b':', ref rest..] => return self.tokenize_symbol(1, Symbol::Colon, rest),
                &[b'#', ref rest..] => return self.tokenize_symbol(1, Symbol::NumberSign, rest),
                &[b'.', ref rest..] => return self.tokenize_symbol(1, Symbol::Dot, rest),
                &[b'*', ref rest..] => return self.tokenize_symbol(1, Symbol::Asterisk, rest),
                &[b'/', ref rest..] => return self.tokenize_symbol(1, Symbol::Slash, rest),
                &[b'+', ref rest..] => return self.tokenize_symbol(1, Symbol::Plus, rest),
                &[b'-', ref rest..] => return self.tokenize_symbol(1, Symbol::Hyphen, rest),
                &[b'<', ref rest..] => return self.tokenize_symbol(1, Symbol::Less, rest),
                &[b'>', ref rest..] => return self.tokenize_symbol(1, Symbol::Greater, rest),
                &[b'!', ref rest..] => return self.tokenize_symbol(1, Symbol::Exclamation, rest),
                &[b'=', ref rest..] => return self.tokenize_symbol(1, Symbol::Equal, rest),
                &[b'?', ref rest..] => return self.tokenize_symbol(1, Symbol::Question, rest),
                &[b'\n', ref rest..] => return self.tokenize_symbol(1, Symbol::NewLine, rest),

                // TODO: handle utf8 correctly
                &[b'$', b'\\', _, ref rest..] => {
                    let position = self.current_position;
                    self.src = rest;
                    self.current_position.forward(3);
                    return Some(Ok(Token {
                        text: &self.src[0..2],
                        position,
                        token_type: TokenType::Integer,
                    }));
                }
                &[b'$', _, ref rest..] => {
                    let position = self.current_position;
                    self.src = rest;
                    self.current_position.forward(2);
                    return Some(Ok(Token {
                        text: &self.src[0..1],
                        position,
                        token_type: TokenType::Integer,
                    }));
                }

                &[b'0'..=b'9', ref rest..] => {
                    let mut size: usize = 1;
                    let mut next = rest;

                    while let &[b'0'..=b'9', ref rest..] = next {
                        size += 1;
                        next = rest;
                    }

                    match next {
                        &[b'#', ref rest..] => {
                            size += 1;
                            next = rest;

                            while let &[b'0'..=b'9', ref rest..] = next {
                                size += 1;
                                next = rest;
                            }

                            let text = &self.src[0..size];
                            let position = self.current_position;

                            self.src = next;
                            self.current_position.forward(size);

                            return Some(Ok(Token {
                                text,
                                position,
                                token_type: TokenType::Integer,
                            }));
                        }
                        _ => panic!(""),
                    }
                }

                &[b'a'..=b'z', ..] | &[b'_', ..] => {
                    let mut size: usize = 0;
                    let mut next = self.src;

                    while let &[b'A'..=b'Z', ref rest..] | &[b'a'..=b'z', ref rest..] | &[b'_', ref rest..] = next {
                        size += 1;
                        next = rest;
                    }

                    let text = &self.src[0..size];
                    let position = self.current_position;

                    self.src = next;
                    self.current_position.forward(size);

                    return Some(Ok(Token {
                        text,
                        position,
                        token_type: TokenType::Atom,
                    }));
                }

                &[b'A'..=b'Z', ..] => {
                    let mut size: usize = 0;
                    let mut next = self.src;

                    while let &[b'A'..=b'Z', ref rest..] | &[b'a'..=b'z', ref rest..] | &[b'_', ref rest..] = next {
                        size += 1;
                        next = rest;
                    }

                    let text = &self.src[0..size];
                    let position = self.current_position;

                    self.src = next;
                    self.current_position.forward(size);

                    return Some(Ok(Token {
                        text,
                        position,
                        token_type: TokenType::Variable,
                    }));
                }

                &[b' ', ref rest..] | &[b'\t', ref rest..] => {
                    self.src = rest;
                    self.current_position.forward(1);
                    continue;
                }

                _ => return None,
            }
        }
    }
}

#[cfg(test)]
#[test]
fn tokenizer_test() {
    assert_debug_snapshot_matches!(
        "tokenize symbols",
        Tokenizer::new(b"( ) , -> { } [ ] | || <- ; : # . \n / + - ++ -- == /= =< < >= \r\n > =:= =/= <= => := << >> ! = :: .. ...")
            .collect::<Result<Vec<_>, _>>()
    );
}
