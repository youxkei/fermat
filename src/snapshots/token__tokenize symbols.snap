---
created: "2019-01-26T11:52:32.215428934+00:00"
creator: insta@0.5.2
expression: "Tokenizer::new(b\"( ) , -> { } [ ] | || <- ; : # . \\n / + - ++ -- == /= =< < >= \\r\\n > =:= =/= <= => := << >> ! = :: .. ...\").collect::<Result<Vec<_>,\n                                                                                                                                              _>>()"
source: src/token.rs
---
Ok(
    [
        Token {
            text: [
                40
            ],
            position: Position {
                offset: 0,
                line: 1,
                column: 1
            },
            token_type: Symbol(
                OpenParen
            )
        },
        Token {
            text: [
                41
            ],
            position: Position {
                offset: 2,
                line: 1,
                column: 3
            },
            token_type: Symbol(
                CloseParen
            )
        },
        Token {
            text: [
                44
            ],
            position: Position {
                offset: 4,
                line: 1,
                column: 5
            },
            token_type: Symbol(
                Comma
            )
        },
        Token {
            text: [
                45,
                62
            ],
            position: Position {
                offset: 6,
                line: 1,
                column: 7
            },
            token_type: Symbol(
                HyphenGreater
            )
        },
        Token {
            text: [
                123
            ],
            position: Position {
                offset: 9,
                line: 1,
                column: 10
            },
            token_type: Symbol(
                OpenBrace
            )
        },
        Token {
            text: [
                125
            ],
            position: Position {
                offset: 11,
                line: 1,
                column: 12
            },
            token_type: Symbol(
                CloseBrace
            )
        },
        Token {
            text: [
                91
            ],
            position: Position {
                offset: 13,
                line: 1,
                column: 14
            },
            token_type: Symbol(
                OpenBracket
            )
        },
        Token {
            text: [
                93
            ],
            position: Position {
                offset: 15,
                line: 1,
                column: 16
            },
            token_type: Symbol(
                CloseBracket
            )
        },
        Token {
            text: [
                124
            ],
            position: Position {
                offset: 17,
                line: 1,
                column: 18
            },
            token_type: Symbol(
                Bar
            )
        },
        Token {
            text: [
                124,
                124
            ],
            position: Position {
                offset: 19,
                line: 1,
                column: 20
            },
            token_type: Symbol(
                BarBar
            )
        },
        Token {
            text: [
                60,
                45
            ],
            position: Position {
                offset: 22,
                line: 1,
                column: 23
            },
            token_type: Symbol(
                LessHyphen
            )
        },
        Token {
            text: [
                59
            ],
            position: Position {
                offset: 25,
                line: 1,
                column: 26
            },
            token_type: Symbol(
                Semicolon
            )
        },
        Token {
            text: [
                58
            ],
            position: Position {
                offset: 27,
                line: 1,
                column: 28
            },
            token_type: Symbol(
                Colon
            )
        },
        Token {
            text: [
                35
            ],
            position: Position {
                offset: 29,
                line: 1,
                column: 30
            },
            token_type: Symbol(
                NumberSign
            )
        },
        Token {
            text: [
                46
            ],
            position: Position {
                offset: 31,
                line: 1,
                column: 32
            },
            token_type: Symbol(
                Dot
            )
        },
        Token {
            text: [
                10
            ],
            position: Position {
                offset: 33,
                line: 1,
                column: 34
            },
            token_type: Symbol(
                NewLine
            )
        },
        Token {
            text: [
                47
            ],
            position: Position {
                offset: 35,
                line: 2,
                column: 2
            },
            token_type: Symbol(
                Slash
            )
        },
        Token {
            text: [
                43
            ],
            position: Position {
                offset: 37,
                line: 2,
                column: 4
            },
            token_type: Symbol(
                Plus
            )
        },
        Token {
            text: [
                45
            ],
            position: Position {
                offset: 39,
                line: 2,
                column: 6
            },
            token_type: Symbol(
                Hyphen
            )
        },
        Token {
            text: [
                43,
                43
            ],
            position: Position {
                offset: 41,
                line: 2,
                column: 8
            },
            token_type: Symbol(
                PlusPlus
            )
        },
        Token {
            text: [
                45,
                45
            ],
            position: Position {
                offset: 44,
                line: 2,
                column: 11
            },
            token_type: Symbol(
                HyphenHyphen
            )
        },
        Token {
            text: [
                61,
                61
            ],
            position: Position {
                offset: 47,
                line: 2,
                column: 14
            },
            token_type: Symbol(
                EqualEqual
            )
        },
        Token {
            text: [
                47,
                61
            ],
            position: Position {
                offset: 50,
                line: 2,
                column: 17
            },
            token_type: Symbol(
                SlashEqual
            )
        },
        Token {
            text: [
                61,
                60
            ],
            position: Position {
                offset: 53,
                line: 2,
                column: 20
            },
            token_type: Symbol(
                EqualLess
            )
        },
        Token {
            text: [
                60
            ],
            position: Position {
                offset: 56,
                line: 2,
                column: 23
            },
            token_type: Symbol(
                Less
            )
        },
        Token {
            text: [
                62,
                61
            ],
            position: Position {
                offset: 58,
                line: 2,
                column: 25
            },
            token_type: Symbol(
                GreaterEqual
            )
        },
        Token {
            text: [
                13,
                10
            ],
            position: Position {
                offset: 61,
                line: 2,
                column: 28
            },
            token_type: Symbol(
                NewLine
            )
        },
        Token {
            text: [
                62
            ],
            position: Position {
                offset: 64,
                line: 3,
                column: 2
            },
            token_type: Symbol(
                Greater
            )
        },
        Token {
            text: [
                61,
                58,
                61
            ],
            position: Position {
                offset: 66,
                line: 3,
                column: 4
            },
            token_type: Symbol(
                EqualColonEqual
            )
        },
        Token {
            text: [
                61,
                47,
                61
            ],
            position: Position {
                offset: 70,
                line: 3,
                column: 8
            },
            token_type: Symbol(
                EqualSlashEqual
            )
        },
        Token {
            text: [
                60,
                61
            ],
            position: Position {
                offset: 74,
                line: 3,
                column: 12
            },
            token_type: Symbol(
                LessEqual
            )
        },
        Token {
            text: [
                61,
                62
            ],
            position: Position {
                offset: 77,
                line: 3,
                column: 15
            },
            token_type: Symbol(
                EqualGreater
            )
        },
        Token {
            text: [
                58,
                61
            ],
            position: Position {
                offset: 80,
                line: 3,
                column: 18
            },
            token_type: Symbol(
                ColonEqual
            )
        },
        Token {
            text: [
                60,
                60
            ],
            position: Position {
                offset: 83,
                line: 3,
                column: 21
            },
            token_type: Symbol(
                LessLess
            )
        },
        Token {
            text: [
                62,
                62
            ],
            position: Position {
                offset: 86,
                line: 3,
                column: 24
            },
            token_type: Symbol(
                GreaterGreater
            )
        },
        Token {
            text: [
                33
            ],
            position: Position {
                offset: 89,
                line: 3,
                column: 27
            },
            token_type: Symbol(
                Exclamation
            )
        },
        Token {
            text: [
                61
            ],
            position: Position {
                offset: 91,
                line: 3,
                column: 29
            },
            token_type: Symbol(
                Equal
            )
        },
        Token {
            text: [
                58,
                58
            ],
            position: Position {
                offset: 93,
                line: 3,
                column: 31
            },
            token_type: Symbol(
                ColonColon
            )
        },
        Token {
            text: [
                46,
                46
            ],
            position: Position {
                offset: 96,
                line: 3,
                column: 34
            },
            token_type: Symbol(
                DotDot
            )
        },
        Token {
            text: [
                46,
                46,
                46
            ],
            position: Position {
                offset: 99,
                line: 3,
                column: 37
            },
            token_type: Symbol(
                DotDotDot
            )
        }
    ]
)
